Documenting binary encoding used with Canal-JSON (#13832)
pingcap/docs: ticdc/ticdc-canal-json.md

For more information about Java SQL Types, see [Java SQL Class Types](https://docs.oracle.com/javase/8/docs/api/java/sql/Types.html).

> ## Binary and Blob types

Reviewer: could you also describe the method of converting a byte array to a string in the documentation？

Author: Are you asking for a worked example in the documentation in a specific programming language like Java or Go? A lot of this would be language specific so not sure what would be useful to further document here.

Here's a minimal go program I wrote to outline the encoding steps taken (encoding the bytes with ISO8859-1, converting to a string, marshaling to JSON) without having to have a full implementation running. Not sure if part of this would address what you are asking for?

```
package main

import (
	"encoding/json"
	"fmt"

	"golang.org/x/text/encoding"
	"golang.org/x/text/encoding/charmap"
)

type Row struct {
	C1 string `json:"c1"`
}

func main() {

	value := []byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221, 222, 223, 224, 225, 226, 227, 228, 229, 230, 231, 232, 233, 234, 235, 236, 237, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 252, 253, 254, 255}

	j1 := encodeIsoJson(value)
	b1 := decodeIsoJson(string(j1))

	fmt.Println("Encoded format")
	fmt.Println(string(j1))
	fmt.Println()
	fmt.Println("Roundtripped bytes")
	fmt.Println(b1)
	fmt.Println()
}

func encodeIsoJson(binarydata []byte) []byte {
	var decoder *encoding.Decoder = charmap.ISO8859_1.NewDecoder()
	decodedBytes, _ := decoder.Bytes(binarydata)
	stringifiedBytes := string(decodedBytes)
	dataRow := &Row{C1: stringifiedBytes}
	jsonDataRow, _ := json.Marshal(dataRow)
	return jsonDataRow
}

func decodeIsoJson(jsonString string) []byte {
	var encoder *encoding.Encoder = charmap.ISO8859_1.NewEncoder()

	jsonData := new(Row)
	json.Unmarshal([]byte(jsonString), jsonData)
	encodedBytes, _ := encoder.Bytes([]byte(jsonData.C1))
	return encodedBytes
}
```

Reviewer: It would be great if there is an example, because customers have given feedback that they do not understand why the canal-json encoding for binary types results in significant data inflation.

Author: Okay, I have added a brief example of the impact of the encoding, including showing how the escaping mechanism adds additional characters. See https://github.com/pingcap/docs/pull/13832/commits/a81c88ee303b94dc375db47ed7167bdcfb4b510b

For more information about Java SQL Types, see [Java SQL Class Types](https://docs.oracle.com/javase/8/docs/api/java/sql/Types.html).

## Binary and Blob types

> TiCDC encodes [binary types](/data-type-string.md##binary-type) in the Canal-JSON format by converting each byte to its character representation using the ISO/IEC 8859-1 character encodings. Non-printable characters are represented using their UTF-8 escape sequence. Certain characters with special meaning in HTML are also escaped using their UTF-8 escape sequence. See the table below for details.

Reviewer 2: ```suggestion
TiCDC encodes [binary types](/data-type-string.md#binary-type) in the Canal-JSON format by converting each byte to its character representation as follows:

- Printable characters are represented using the ISO/IEC 8859-1 character encodings. 
- Non-printable characters and certain characters with special meaning in HTML are represented using their UTF-8 escape sequence. 

The following table shows the detailed representation information.
```

> | Description                 | Value range | String representation |

Reviewer 2: What does "Description" in the header represent? The "Character type"?

> | Description                 | Value range | String representation |

Reviewer 2: ```suggestion
| Character type                | Value range | Character representation |
```

| :---------------------------| :-----------| :---------------------|

> | Control characters          | [0, 31]     | UTF-8 escape, e.g. '\u0000' through '\u001F' |

Reviewer 2: ```suggestion
| Control characters          | [0, 31]     | UTF-8 escape (such as '\u0000' through '\u001F') |
```

> | Horizontal tab              | [9]         | \t                    |
> | Line feed                   | [10]        | \n                    |
> | Cariage return              | [13]        | \r                    |

Reviewer 2: ```suggestion
| Horizontal tab              | [9]         | `\t`                    |
| Line feed                   | [10]        | `\n`                    |
| Cariage return              | [13]        | `\r`                    |
```

> | Printable characters        | [32, 127]   | Literal character (e.g. 'A') |

Reviewer 2: ```suggestion
| Printable characters        | [32, 127]   | Literal character (such as 'A') |
```

> | Ampersand                   | [38]        | \u0026                |
> | Less-than sign              | [60]        | \u0038                |
> | Greater-than sign           | [62]        | \u003E                |

Reviewer 2: ```suggestion
| Ampersand                   | [38]        | `\u0026`                |
| Less-than sign              | [60]        | `\u0038`                |
| Greater-than sign           | [62]        | `\u003E`                |
```

| Extended control characters | [128, 159]  | Literal character   |
| ISO 8859-1 (Latin-1)        | [160, 255]  | Literal character   |

### Example of the encoding

> To illustrate, the following 16 bytes `[5 7 10 15 36 50 43 99 120 60 38 255 254 45 55 70]` stored in a `VARBINARY` called `c_varbinary` would be encoded in a Canal-JSON `Update` Event as:

Reviewer 2: ```suggestion
For example, the following 16 bytes `[5 7 10 15 36 50 43 99 120 60 38 255 254 45 55 70]` stored in a `VARBINARY` column called `c_varbinary` is encoded in a Canal-JSON `Update` event as:
```

For more information about Java SQL Types, see [Java SQL Class Types](https://docs.oracle.com/javase/8/docs/api/java/sql/Types.html).

## Binary and Blob types

TiCDC encodes [binary types](/data-type-string.md#binary-type) in the Canal-JSON format by converting each byte to its character representation as follows:

- Printable characters are represented using the ISO/IEC 8859-1 character encodings. 
- Non-printable characters and certain characters with special meaning in HTML are represented using their UTF-8 escape sequence. 

The following table shows the detailed representation information.

| Character type                | Value range | Character representation |
| :---------------------------| :-----------| :---------------------|

> | Control characters          | [0, 31]     | UTF-8 escape (such as '\u0000' through '\u001F') |

Reviewer 2: ```suggestion
| Control characters          | [0, 31]     | UTF-8 escape (such as `\u0000` through `\u001F`) |
```

| Horizontal tab              | [9]         | `\t`                    |
| Line feed                   | [10]        | `\n`                    |

> | Cariage return              | [13]        | `\r`                    |

Reviewer 3: ```suggestion
| Carriage return              | [13]        | `\r`                    |
```

Author: Addressed in [90f3244](https://github.com/pingcap/docs/pull/13832/commits/90f324453ed9a0f8973bd08b2afae26a35852857)

> | Printable characters        | [32, 127]   | Literal character (such as 'A') |

Reviewer 2: ```suggestion
| Printable characters        | [32, 127]   | Literal character (such as `A`) |
```

| Ampersand                   | [38]        | `\u0026`                |
| Less-than sign              | [60]        | `\u0038`                |
| Greater-than sign           | [62]        | `\u003E`                |
| Extended control characters | [128, 159]  | Literal character   |
| ISO 8859-1 (Latin-1)        | [160, 255]  | Literal character   |

### Example of the encoding

> For example, the following 16 bytes `[5 7 10 15 36 50 43 99 120 60 38 255 254 45 55 70]` stored in a `VARBINARY` column called `c_varbinary` is encoded in a Canal-JSON `Update` event as:

Reviewer 3: ```suggestion
For example, the following 16 bytes `[5 7 10 15 36 50 43 99 120 60 38 255 254 45 55 70]` stored in a `VARBINARY` column called `c_varbinary` are encoded in a Canal-JSON `Update` event as follows:
```

Author: Addressed in [90f3244](https://github.com/pingcap/docs/pull/13832/commits/90f324453ed9a0f8973bd08b2afae26a35852857)
